File: java/CombinationsOfLength.java
import java.util.LinkedList;
import java.util.List;

public class CombinationsOfLength {
    public static List<List<Integer>> combinationsOfLengthBruteForce(int[] n, int length) {
        List<List<Integer>> allCombinations = Combinations.combinationsBuiltUp(n);
        
        List<List<Integer>> filtered = new LinkedList<List<Integer>>();
        for (List<Integer> l : allCombinations) {
            if (l.size() == length) filtered.add(l);
        }
        
        return filtered;
    }
    
    public static List<List<Integer>> combinationsOfLengthBacktracking(int[] n, int length) {
        return combinationsOfLengthBacktracking(n, 0, length, 0);
    }
    
    private static List<List<Integer>> combinationsOfLengthBacktracking(int[] n, int i, 
                                                                       int targetLength, 
                                                                       int currentLength) {
        if (currentLength > targetLength) return new LinkedList<List<Integer>>();
        if (i == n.length && currentLength != targetLength) return new LinkedList<List<Integer>>();
        if (i == n.length) {
            List<List<Integer>> toReturn = new LinkedList<List<Integer>>();
            toReturn.add(new LinkedList<Integer>());
            return toReturn;
        }
        
        List<List<Integer>> include = combinationsOfLengthBacktracking(n, i+1, targetLength, currentLength + 1);
        List<List<Integer>> exclude = combinationsOfLengthBacktracking(n, i+1, targetLength, currentLength);
        
        List<List<Integer>> toReturn = new LinkedList<List<Integer>>();
        for (List<Integer> result : include) {
            result.add(0, n[i]);
            toReturn.add(result);
        }
        
        toReturn.addAll(exclude);
        
        return toReturn;
    }
    
//        private static List<List<Integer>> combinationsBuiltUp(int[] n, int i) {
//        if (i == n.length) {
//            List<List<Integer>> toReturn = new LinkedList<List<Integer>>();
//            toReturn.add(new LinkedList<Integer>());
//            return toReturn;
//        }
//        
//        List<List<Integer>> toReturn = new LinkedList<List<Integer>>();
//        for (List<Integer> result : combinationsBuiltUp(n, i+1)) {
//            // Exclude current item
//            toReturn.add(new LinkedList<Integer>(result));
//            // Include current item
//            result.add(0, n[i]);
//            toReturn.add(new LinkedList<Integer>(result));
//        }
//        
//        return toReturn;
//    }
    
    public static void main(String[] args) {
        System.out.println(combinationsOfLengthBruteForce(new int[]{1,2,3,4,5}, 3));
        System.out.println(combinationsOfLengthBacktracking(new int[]{1,2,3,4,5}, 3));
    }
}

File: python3/CombinationsOfLength.py
class CombinationsOfLength():
	def combinations_of_length_brute_force(self, n, length):
		# the below method is from Combinations.py file
		all_combinations = Cominations.combinations_built_up(n)
		filtered = []
		for l in all_combinations:
			if len(l) == length:
				filtered.append(l)
		return filtered

	def combinations_of_length_backtracking(self, n, length):
		return self.combinations_of_length_backtracking_helper(n, 0, length, 0)

	def combinations_of_length_backtracking_helper(self, n, i, target_length, current_length):
		if current_length > target_length:
			return []

		if i == len(n) and current_length != target_length:
			return []

		if i == len(n):
			to_return = []
			to_return.append([])
			return to_return

		include = self.combinations_of_length_backtracking_helper(n, i + 1, target_length, current_length + 1)
		exclude = self.combinations_of_length_backtracking_helper(n, i + 1, target_length, current_length)

		to_return = []

		for result in include:
			result.insert(0, n[i])
			to_return.append(result)

		to_return.extend(exclude)

		return to_return

test = CombinationsOfLength()
print(test.combinations_of_length_backtracking([1,2,3,4,5], 3))
